src/lib/kdf/hkdf/hkdf.cpp

```diff
@@ -93,75 +80,54 @@ std::string HKDF_Expand::name() const {
+   if(key.empty()) {
       return;
    }
 
-   if(key_len > m_prf->output_length() * 255) {
-      throw Invalid_Argument("HKDF-Expand maximum output length exceeeded");
-   }
-
-   m_prf->set_key(secret, secret_len);
+   BOTAN_ARG_CHECK(key.size() <= m_prf->output_length() * 255, "HKDF-Expand maximum output length exceeeded");
 
-   uint8_t counter = 1;
    secure_vector<uint8_t> h;
-   size_t offset = 0;
 
-   while(offset != key_len) {
+   BufferStuffer k(key);
+   m_prf->set_key(secret);
+   for(uint8_t counter = 1; !k.full(); ++counter) {
       m_prf->update(h);
-      m_prf->update(label, label_len);
-      m_prf->update(salt, salt_len);
-      m_prf->update(counter++);
+      m_prf->update(label);
+      m_prf->update(salt);
+      m_prf->update(counter);
       m_prf->final(h);
 
-      const size_t written = std::min(h.size(), key_len - offset);
-      copy_mem(&key[offset], h.data(), written);
-      offset += written;
+      const auto bytes_to_write = std::min(h.size(), k.remaining_capacity());
+      k.append(std::span{h}.first(bytes_to_write));
    }
 }
 
 secure_vector<uint8_t> hkdf_expand_label(std::string_view hash_fn,
-                                         const uint8_t secret[],
-                                         size_t secret_len,
+                                         std::span<const uint8_t> secret,
                                          std::string_view label,
-                                         const uint8_t hash_val[],
-                                         size_t hash_val_len,
+                                         std::span<const uint8_t> hash_val,
                                          size_t length) {
    BOTAN_ARG_CHECK(length <= 0xFFFF, "HKDF-Expand-Label requested output too large");
    BOTAN_ARG_CHECK(label.size() <= 0xFF, "HKDF-Expand-Label label too long");
-   BOTAN_ARG_CHECK(hash_val_len <= 0xFF, "HKDF-Expand-Label hash too long");
-
-   const uint16_t length16 = static_cast<uint16_t>(length);
+   BOTAN_ARG_CHECK(hash_val.size() <= 0xFF, "HKDF-Expand-Label hash too long");
 
    HKDF_Expand hkdf(MessageAuthenticationCode::create_or_throw(fmt("HMAC({})", hash_fn)));
 
-   secure_vector<uint8_t> output(length16);
-   std::vector<uint8_t> prefix(3 + label.size() + 1);
-
-   prefix[0] = get_byte<0>(length16);
-   prefix[1] = get_byte<1>(length16);
-   prefix[2] = static_cast<uint8_t>(label.size());
-
-   copy_mem(prefix.data() + 3, cast_char_ptr_to_uint8(label.data()), label.size());
-
-   prefix[3 + label.size()] = static_cast<uint8_t>(hash_val_len);
+   const auto prefix = concat<std::vector<uint8_t>>(store_be(static_cast<uint16_t>(length)),
+                                                    store_be(static_cast<uint8_t>(label.size())),
+                                                    std::span{cast_char_ptr_to_uint8(label.data()), label.size()},
+                                                    store_be(static_cast<uint8_t>(hash_val.size())));
```

Comment:
This is *probably* better overall than what's currently in master, but also I hate it :)

I think what I want for things like this is some kind of "binary format string" eg

> const auto prefix = format_binary("%u16be %u8bytes %u8", length, label, hash_val);

(not exactly this, those format string choices suck and are just for exposition; there is probably a lot of prior art here)

Anyway no change required here.

Reply:
Agree. That would be really helpful. Perhaps even in the public API (at some point). I'd love to have something like that on application level as well. :)